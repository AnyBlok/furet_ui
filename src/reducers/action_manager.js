/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
export const defaultState = {
    actions: [],
    action_data: {},
}

export const action_manager = (state = defaultState, action) => {
    const value = Object.assign({}, action);
    delete value.type;
    switch (action.type) {
        case 'UPDATE_ACTION_MANAGER_ADD_ACTION':
            const actions = state.actions.slice(0);
            actions.push(action.actionId);
            return Object.assign({}, state, {actions});
        case 'UPDATE_ACTION_MANAGER_ADD_ACTION_DATA':
            delete value.actionId;
            const action_data = Object.assign({}, state.action_data);
            action_data[action.actionId] = value;
            return Object.assign({}, state, {action_data});
        case 'RESET_ACTION_MANAGER':
            return Object.assign({}, state, {actions: []});
        case 'CLEAR_ACTION_MANAGER':
            return defaultState;
        default:
            return state
    }
}

export default action_manager;
