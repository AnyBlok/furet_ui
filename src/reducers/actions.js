const defaultState = {}

export const actions = (state = defaultState, action) => {
    const value = Object.assign({}, action);
    delete value.type;
    switch (action.type) {
        case 'UPDATE_ACTION_SELECT_VIEW':
            delete value.actionId;
            const values = {};
            values[action.actionId] = Object.assign({}, state[action.actionId], value);
            return Object.assign({}, state, values);
        case 'CLEAR_ACTION':
            return defaultState;
        default:
            return state
    }
}

export default actions;
