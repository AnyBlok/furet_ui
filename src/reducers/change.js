/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
export const defaultState = {
    current: {},
    toSync: [],
    computed: {},
    currents: {},
}

export const current2Sync = (current, toSync, uuid, model, dataId, newData, fields) => {
    const newSync = {
        uuid,
        state: 'toSend',
        data: [],
    }
    // Add main data in first
    try {
        newSync.data.push({
            model,
            dataId,
            type: newData ? 'CREATE' : 'UPDATE',
            fields,
            data: current[model][dataId],
        });
    } catch (e) {};
    // Add other data
    _.each(_.keys(current), modelName => {
        _.each(_.keys(current[modelName]), id => {
            if (modelName == model && dataId == id) {
                // do nothing, because already done
            } else {
                const data = current[modelName][id];
                if (data == 'DELETED'){
                    newSync.data.push({
                        model: modelName,
                        dataIds: [id],
                        type: 'DELETE',
                    });
                }else {
                    const _fields = current[modelName][id].__fields;
                    delete current[modelName][id].__fields;
                    newSync.data.push({
                        model: modelName,
                        dataId: id,
                        type: (new RegExp('^new-.*')).test(id) ? 'CREATE' : 'UPDATE',
                        fields: _fields,
                        data,
                    });
                }
            }
        });
    });
    if (newSync.data.length > 0) {
        toSync.push(newSync);
    }
}

export const remove2Sync = (toSync, uuid, model, dataIds) => {
    if (dataIds.length == 0) return;
    toSync.push({
        uuid,
        state: 'toSend',
        data: [
            {
                model,
                dataIds,
                type: 'DELETE',
            }
        ],
    });
}

export const toSync2computed = (toSync, computed) => {
    _.each(toSync, ts => {
        _.each(ts.data, data => {
            if (computed[data.model] == undefined) computed[data.model] = {};
            switch (data.type) {
                case 'CREATE':
                case 'UPDATE':
                    if (computed[data.model][data.dataId] == undefined) computed[data.model][data.dataId] = {};
                    Object.assign(computed[data.model][data.dataId], data.data);
                    break;
                case 'DELETE':
                    _.each(data.dataIds, dId => {
                        computed[data.model][dId] = 'DELETED';
                    });
                    break;
            }
        });
    });
}

export const removeUuid = (toSync, uuid) => {
    return _.filter(toSync, s => s.uuid != uuid);
}

export const changeState = (toSync, uuid, state) => {
    return _.map(toSync, s => {
        const val = Object.assign({}, s)
        if (uuid == s.uuid) {
            val.state = state;
        }
        return val;
    });
}

export const change = (state = defaultState, action) => {
    const current = Object.assign({}, state.current),
          toSync = state.toSync.slice(0),
          computed = {},
          currents = Object.assign({}, state.currents);
    switch (action.type) {
        case 'ON_CHANGE':
            if (current[action.model] == undefined) current[action.model] = {};
            else current[action.model] = Object.assign({}, current[action.model]);
            if (current[action.model][action.dataId] == undefined) current[action.model][action.dataId] = {};
            else current[action.model][action.dataId] = Object.assign({}, current[action.model][action.dataId]);
            current[action.model][action.dataId][action.fieldname] = action.newValue;
            if (action.fields) {
                current[action.model][action.dataId]['__fields'] = action.fields;
            }
            return Object.assign({}, state, {current});
        case 'ON_CHANGE_DELETE':
            if (current[action.model] == undefined) current[action.model] = {};
            else current[action.model] = Object.assign({}, current[action.model]);
            _.each(action.dataIds, dataId => {
                current[action.model][dataId] = 'DELETED'
            });
            return Object.assign({}, state, {current});
        case 'CLEAR_CHANGE':
            return Object.assign({}, state, {current: {}});
        case 'ON_SAVE':
            current2Sync(current, toSync, action.uuid, action.model, action.dataId, action.newData, action.fields);
            toSync2computed(toSync, computed)
            return Object.assign({}, state, {current: {}, toSync, computed});
        case 'ON_DELETE':
            remove2Sync(toSync, action.uuid, action.model, action.dataIds);
            toSync2computed(toSync, computed)
            return Object.assign({}, state, {current: {}, toSync, computed});
        case 'SYNC':
            return Object.assign({}, state, {toSync: changeState(toSync, action.uuid, 'Sent')});
        case 'UNSYNC':
            return Object.assign({}, state, {toSync: changeState(toSync, action.uuid, 'toSend')});
        case 'SYNCED':
            let _toSync = removeUuid(toSync, action.uuid);
            toSync2computed(_toSync, computed)
            return Object.assign({}, state, {toSync: _toSync, computed});
        case 'ADD_CURRENTS':
            const c = {};
            c[action.actionId] = current;
            Object.assign(currents, c);
            return Object.assign({}, state, {current: {}, currents});
        case 'REVERT_CHANGE':
            const prev_current = currents[action.actionId];
            delete currents[action.actionId];
            _.each(action.actionIds, actionId => {
                delete currents[actionId];
            });
            return Object.assign({}, state, {current: prev_current, currents});
        default:
            return state
    }
}

export default change;
