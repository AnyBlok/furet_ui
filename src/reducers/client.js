const defaultState = {}

export const client = (state = defaultState, action) => {
    const value = Object.assign({}, action);
    delete value.type;
    switch (action.type) {
        case 'UPDATE_VIEW_CLIENT':
            delete value.viewName;
            const values = {};
            values[action.viewName] = value
            return Object.assign({}, state, values);
        case 'CLEAR_CLIENT':
            return defaultState;
        default:
            return state
    }
}

export default client;

