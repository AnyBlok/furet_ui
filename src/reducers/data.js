/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
export const defaultState = {}

export const data = (state = defaultState, action) => {
    switch (action.type) {
        case 'UPDATE_DATA':
            const values = Object.assign({}, state);
            if (values[action.model] == undefined) values[action.model] = {};
            else values[action.model] = Object.assign({}, values[action.model]);
            Object.assign(values[action.model], action.data);
            return values;
        case 'CLEAR_DATA':
            return defaultState;
        default:
            return state
    }
}

export default data;

