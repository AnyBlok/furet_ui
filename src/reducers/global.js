/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
export const defaultState = {
    title: '',
    custom_view: '',
    modal_custom_view: '',
    spaceId: '',
}

export const global = (state = defaultState, action) => {
    const value = Object.assign({}, action);
    delete value.type;
    switch (action.type) {
        case 'UPDATE_GLOBAL':
            const _state = Object.assign({}, defaultState);
            if (state.title) {
                delete _state.title;
                delete _state.modal_custom_view;
            }
            if (action.custom_view || action.spaceId) return Object.assign({}, state, _state, value);
            else return Object.assign({}, state, value);
        case 'CLEAR_GLOBAL':
            return defaultState;
        default:
            return state
    }
}

export default global;
