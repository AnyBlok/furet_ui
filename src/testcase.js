import _ from 'underscore';

export const updateGlobal = () => {
    global.document = {
        createElement: () => {
            return {
                innerHTM: () => {},
            };
        },
        body: {
            appendChild: () => {},
        }
    };
    global.window = {
        addEventListener: () => {},
        attachEvent: () => {},
    }
    global.navigator = global.navigator || {};
    global.navigator.userAgent = 'all';
    global._ = _;
}

export default {
    updateGlobal,
}
