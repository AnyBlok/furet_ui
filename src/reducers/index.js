/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
import global from './global';
import client from './client';
import rightmenu from './right-menu';
import leftmenu from './left-menu';
import spaces from './spaces';
import {action_manager} from './action_manager';
import actions from './actions';
import views from './views';
import data from './data';
import translate from 'counterpart';
import * as moment from 'moment';
import 'moment';


export const dispatchAll = (dispatch, datas) => {
    _.each(datas, data => {
        if (data.type) {
            switch (data.type) {
                case 'UPDATE_LOCALES':
                    _.each(data.locales, d => {
                        translate.registerTranslations(d.locale, {furetUI: d.counterpart});
                    });
                    break
                case 'SET_LOCALE':
                    translate.setLocale(data.locale)
                    moment.locale(data.locale) 
                default:
                    dispatch(data);
            }
        }
    });
}

export default {
    global,
    client,
    rightmenu,
    leftmenu,
    spaces,
    action_manager,
    actions,
    views,
    data,
}
