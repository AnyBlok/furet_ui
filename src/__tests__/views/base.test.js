/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import React from 'react';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';
import reducers from '../../reducers';
import {updateGlobal} from '../../testcase';
import BaseView, {renderSafeEval} from '../../views/base';
import chai from 'chai';

jest.mock('../../server-call')

test('Render Base View', () => {
    const store = createStore(combineReducers(reducers));
    updateGlobal();
    const component = renderer.create(
        <Provider store={store}>
            <MuiThemeProvider>
                <BaseView/>
            </MuiThemeProvider>
        </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Check renderSafeEval undefined', () => {
    chai.expect(renderSafeEval(undefined, {})).to.equal(false);
});
test('Check renderSafeEval true', () => {
    chai.expect(renderSafeEval('true', {})).to.equal(true);
});
test('Check renderSafeEval 1', () => {
    chai.expect(renderSafeEval('1', {})).to.equal(true);
});
test('Check renderSafeEval fields.b == 1', () => {
    chai.expect(renderSafeEval('fields.b == 1', {b: 1})).to.equal(true);
});
test('Check renderSafeEval fields.b != 1', () => {
    chai.expect(renderSafeEval('fields.b != 1', {b: 1})).to.equal(false);
});
test('Check renderSafeEval toDate(fields.b) != now', () => {
    chai.expect(renderSafeEval('toDate(fields.b) != now', {b: '2017-05-09T07:04:49'})).to.equal(true);
});
