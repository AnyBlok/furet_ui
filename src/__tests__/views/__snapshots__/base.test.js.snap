// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Login 1`] = `
<div>
  <div
    className="row"
  />
  Base view
</div>
`;
